File: Hash Tables/ht-chaining.java
// Hashing Collision Resolution - Chaining
// Avg Time: O(1), Worst: O(n)
// Space: O(n)

import java.util.LinkedList;

class HashNode {
    int key, value;
    HashNode(int k, int v) { key = k; value = v; }
}

public class HashTableChaining {
    private int size = 10;
    private LinkedList<HashNode>[] table;

    public HashTableChaining() {
        table = new LinkedList[size];
        for (int i = 0; i < size; i++) table[i] = new LinkedList<>();
    }

    private int hashFunction(int key) {
        return key % size;
    }

    public void insert(int key, int value) {
        int index = hashFunction(key);
        for (HashNode node : table[index]) {
            if (node.key == key) {
                node.value = value;
                return;
            }
        }
        table[index].add(new HashNode(key, value));
    }

    public Integer search(int key) {
        int index = hashFunction(key);
        for (HashNode node : table[index]) {
            if (node.key == key) return node.value;
        }
        return null;
    }

    public void delete(int key) {
        int index = hashFunction(key);
        table[index].removeIf(node -> node.key == key);
    }
}

File: Hash Tables/ht-chaining.py
# Hashing Collision Resolution - Chaining
# Average Time: O(1), Worst: O(n)
# Space: O(n)

class HashTableChaining:
    def __init__(self, size=10):
        self.size = size
        self.table = [[] for _ in range(size)]

    def hash_function(self, key):
        return key % self.size

    def insert(self, key, value):
        index = self.hash_function(key)
        for i, (k, _) in enumerate(self.table[index]):
            if k == key:
                self.table[index][i] = (key, value)
                return
        self.table[index].append((key, value))

    def search(self, key):
        index = self.hash_function(key)
        for k, v in self.table[index]:
            if k == key:
                return v
        return None

    def delete(self, key):
        index = self.hash_function(key)
        for i, (k, _) in enumerate(self.table[index]):
            if k == key:
                self.table[index].pop(i)
                return
